plugins/auth-backend/src/providers/gcp-iap/provider.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AuthProviderRouteHandlers,
+  AuthProviderFactoryOptions,
+  ExperimentalIdentityResolver,
+} from '../types';
+
+import express from 'express';
+
+import { Logger } from 'winston';
+import { CatalogApi } from '@backstage/catalog-client';
+
+import { OAuth2Client } from 'google-auth-library';
+
+const IAP_JWT_HEADER = 'x-goog-iap-jwt-assertion';
+
+export type GcpIAPProviderOptions = {
+  audience: string;
+  identityResolutionCallback: ExperimentalIdentityResolver;
+};
+export class GcpIAPProvider implements AuthProviderRouteHandlers {
+  private logger: Logger;
+  private options: GcpIAPProviderOptions;
+  private readonly catalogClient: CatalogApi;
+
+  constructor(
+    logger: Logger,
+    catalogClient: CatalogApi,
+    options: GcpIAPProviderOptions,
+  ) {
+    this.logger = logger;
+    this.catalogClient = catalogClient;
+    this.options = options;
+  }
+  frameHandler(): Promise<void> {
+    return Promise.resolve(undefined);
+  }
+
+  async refresh(req: express.Request, res: express.Response): Promise<void> {
+    const expectedAudience = this.options.audience;
+    const jwtToken = req.header(IAP_JWT_HEADER);
+    if (jwtToken === undefined) {
+      res.status(401);
+      res.end();
+      return;
+    }
+    const oAuth2Client = new OAuth2Client();
+    const verify = async () => {
+      const response = await oAuth2Client.getIapPublicKeys();
+      const ticket = await oAuth2Client.verifySignedJwtWithCertsAsync(
+        jwtToken,
+        response.pubkeys,
+        expectedAudience,
+        ['https://cloud.google.com/iap'],
+      );
+      return ticket.getPayload();
+    };
+
+    try {
+      const user = await verify();
+      if (user === undefined) {
+        this.logger.error('gcp iap proxy user returned undefined');
+        res.status(401);
```

Comment:
Let's just throw errors instead so that we leave the error handling more flexible